cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -17,32 +17,37 @@
 package co.cask.cdap.internal.app.runtime.workflow;
 
 import co.cask.cdap.api.workflow.NodeValueEntry;
+import co.cask.cdap.api.workflow.Value;
 import co.cask.cdap.api.workflow.WorkflowToken;
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 
+import java.util.EnumMap;
 import java.util.List;
 import java.util.Map;
-import javax.annotation.Nullable;
 
 /**
  * Implementation of the {@link WorkflowToken} interface.
  */
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
-  private final Map<String, List<NodeValueEntry>> tokenValueMap = Maps.newHashMap();
+  private final Map<Scope, Map<String, List<NodeValueEntry>>> tokenValueMap = new EnumMap<>(Scope.class);
   private String nodeName;
 
   public BasicWorkflowToken() {
   }
 
-  public BasicWorkflowToken(BasicWorkflowToken other) {
-    for (Map.Entry<String, List<NodeValueEntry>> entry : other.tokenValueMap.entrySet()) {
-      List<NodeValueEntry> nodeValueList = Lists.newArrayList();
-      nodeValueList.addAll(entry.getValue());
-      this.tokenValueMap.put(entry.getKey(), nodeValueList);
+  private BasicWorkflowToken(BasicWorkflowToken other) {
+    for (Map.Entry<Scope, Map<String, List<NodeValueEntry>>> entry : other.tokenValueMap.entrySet()) {
+      Map<String, List<NodeValueEntry>> tokenValueMapByScope = Maps.newHashMap();
```

Comment:
also nit: ``tokenValueMapForScope``, since this is for a single scope. 'by' suggests that this data structure contains multiple token values keyed by scope